report/sections/Frontend/chapter.tex

```diff
@@ -0,0 +1,86 @@
+
+\begin{figure}[H]
+    \centering
+    \frame{\includegraphics[scale=0.1]{home.png}}
+    \caption{Home page as a logged-in user}
+    \label{fig:Home page}
+\end{figure}
+
+\section{Syllabi, Problem sets and Exercises}
+On the website, each page that displays course data makes a request to a Next.js backend API endpoint to retrieve the data from the database.
+This request is made using tRPC, which allows for efficient communication between the frontend and the backend. tRPC wraps around React Query, which provides useful features such as caching and automatic refetching.
+
+The request input is validated using zod to ensure that it is in the correct format.
+The user's authorization is also checked to ensure that they are allowed to access the data they have requested.
+Once the input has been validated, the database is queried using Prisma to retrieve the requested data.
+
+The actions that a user is allowed to take on a page depend on their role. For example, a lecturer may have the ability to create, edit, and delete syllabi, problem sets, and exercises, whereas a student may only be able to read the information on the page.
+
+There is a modal on the page that contains form controls for creating new elements, such as syllabi or problem sets.
+This modal helps to prevent invalid input by ensuring that the user enters the correct information before creating a new element.
+Overall, our website supports CRUD actions for lecturers and read-only access for students.
+
+\begin{figure}[H]
+    \centering
+    \frame{\includegraphics[scale=0.1]{syllabi.png}}
+    \caption{Syllabi overview as a logged-in student}
+    \label{fig:syllabi}
+\end{figure}
+
+\begin{figure}[H]
+    \centering
+    \frame{\includegraphics[scale=0.1]{problemsets.png}}
+    \caption{Problem sets overview as a logged-in student}
+    \label{fig:problemsets}
+\end{figure}
+
+\begin{figure}[H]
+    \centering
+    \frame{\includegraphics[scale=0.1]{exercises.png}}
+    \caption{Exercise overview}
+    \label{fig:exercise_overview}
+\end{figure}
+
+\begin{figure}[H]
+    \centering
+    \frame{\includegraphics[scale=0.1]{problemsets_teacher_view.png}}
+    \caption{Problem sets overview as a logged-in lecturer}
+    \label{fig:problemsets_teacher_view}
+\end{figure}
+
+\section{Problem View and Code Editor}
+When a user visits a page on our website, we fetch the code template exercise on the server in order to prevent flashing in the template after the page has loaded.
+This is important because, without this, users with slow internet connections might see the template being overwritten after they have already started writing in it. That is, if you don't disable the input field entirely until the page has finished loading data. Both these approaches could lead to confusion and frustration for the user, so we decided to fetch the template before the page is rendered.
```

Comment:
```suggestion
This is important because, without this, users with slow internet connections might see the template being overwritten after they have already started writing in it.  These approaches could lead to confusion and frustration for the user, so we decided to fetch the template before the page is rendered.
```

Reply:
Want to show that we've considered other options, not just called it here.